lib/web_ui/lib/src/engine/canvaskit/noto_font.dart

```diff
@@ -0,0 +1,85 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+class NotoFont {
+  const NotoFont(this.name, this.url, this._rangeStarts, this._rangeEnds);
+
+  factory NotoFont.fromFlatRanges(String name, String url, List<int> flatRanges) {
+    final List<int> starts = <int>[];
+    final List<int> ends = <int>[];
+    for (int i = 0; i < flatRanges.length; i += 2) {
+      starts.add(flatRanges[i]);
+      ends.add(flatRanges[i+1]);
+    }
+    return NotoFont(name, url, starts, ends);
+  }
+
+  final String name;
+  final String url;
+  // A sorted list of Unicode range start points.
+  final List<int> _rangeStarts;
+
+  // A sorted list of Unicode range end points.
+  final List<int> _rangeEnds;
+
+  List<CodeunitRange> get unicodeRanges {
+    final List<CodeunitRange> result = <CodeunitRange>[];
+    for (int i = 0; i < _rangeStarts.length; i++) {
+      result.add(CodeunitRange(_rangeStarts[i], _rangeEnds[i]));
+    }
+    return result;
+  }
+
+  // Returns `true` if this font has a glyph for the given [codeunit].
+  bool contains(int codeUnit) {
+    // Binary search through the starts and ends to see if there
+    // is a range that contains the codeunit.
+    int min = 0;
+    int max = _rangeStarts.length;
+    // Search for the first rangeStart that is greater than codeunit.
+    while (min < max) {
+      final int mid = (min + max) ~/ 2;
+      if (_rangeStarts[mid] > codeUnit) {
+        if (mid == 0) return false;
+        final int rangeStart = _rangeStarts[mid - 1];
+        if (rangeStart <= codeUnit) {
+          final int rangeEnd = _rangeEnds[mid - 1];
+          return rangeStart <= codeUnit && codeUnit <= rangeEnd;
+        } else {
+          max = mid - 1;
+        }
+      } else if (_rangeStarts[mid] < codeUnit) {
+        min = mid;
+      }
+    }
+    return false;
```

Comment:
Looks like the while loop never considers the last range (because there's no rangeStart that is greater that codeunit after the last range), so before returning `false` this might need to check if the last range contains the code unit.

Reply:
Done